taichi/backends/metal/struct_metal.cpp

```diff
@@ -340,15 +336,39 @@ class StructCompiler {
     return sn_desc.stride;
   }
 
+  template <typename... Args>
+  void emit(std::string f, Args &&... args) {
+    line_appender_.append(std::move(f), std::move(args)...);
+  }
+
+  std::vector<SNode *> snodes_;
+  int max_snodes_{0};
+  LineAppender line_appender_;
+  std::unordered_map<int, SNodeDescriptor> snode_descriptors_;
+  bool has_sparse_snode_{false};
+};
+
+class RuntimeModuleCompiler {
+ public:
+  CompiledRuntimeModule run() {
+    CompiledRuntimeModule res;
+    emit_runtime_structs();
+    line_appender_.dump(&res.runtime_utils_source_code);
+    res.rand_seeds_size = compute_rand_seeds_size();
+    res.runtime_size = compute_snodes_runtime_size() + res.rand_seeds_size;
+    return res;
+  }
+
+ private:
   void emit_runtime_structs() {
     line_appender_.append_raw(shaders::kMetalRuntimeStructsSourceCode);
     emit("");
     emit("struct Runtime {{");
-    emit("  SNodeMeta snode_metas[{}];", max_snodes_);
-    emit("  SNodeExtractors snode_extractors[{}];", max_snodes_);
-    emit("  ListManagerData snode_lists[{}];", max_snodes_);
-    emit("  NodeManagerData snode_allocators[{}];", max_snodes_);
-    emit("  NodeManagerData::ElemIndex ambient_indices[{}];", max_snodes_);
+    emit("  SNodeMeta snode_metas[{}];", kMaxNumSNodes);
```

Comment:
How big is kMaxNumSNodes? Will it be a concern in the future

Reply:
Good point! This is 1024 now https://github.com/taichi-dev/taichi/blob/fce9108c7c3cfda1b3b828212606f38ded7144ba/taichi/inc/constants.h#L12

Unfortunately the sparse runtime part cannot be easily separated into a per-root buffer yet, so all the sparse info (for multiple roots) still share a single global buffer.